### PR TITLE
SQS-375 | Unit test createFormattedLimitOrder

### PR DESCRIPTION
**Pull request by deividaspetraitis**
This PR introduces unit tests for Orderbook usecase `createFormattedLimitOrder` method. The aim of the tests are to cover edge cases as well as successful scenario. As a side effect pull request introduces `OrderbookRepositoryMock`.

```
❯ go test ./orderbook/usecase -run "TestOrderbookUsecaseTestSuite/TestCreateFormattedLimitOrder" -v
=== RUN   TestOrderbookUsecaseTestSuite
=== RUN   TestOrderbookUsecaseTestSuite/TestCreateFormattedLimitOrder
=== RUN   TestOrderbookUsecaseTestSuite/TestCreateFormattedLimitOrder/tick_not_found
=== RUN   TestOrderbookUsecaseTestSuite/TestCreateFormattedLimitOrder/error_parsing_quantity
=== RUN   TestOrderbookUsecaseTestSuite/TestCreateFormattedLimitOrder/error_parsing_placed_quantity
=== RUN   TestOrderbookUsecaseTestSuite/TestCreateFormattedLimitOrder/error_getting_spot_price_scaling_factor
=== RUN   TestOrderbookUsecaseTestSuite/TestCreateFormattedLimitOrder/error_parsing_bid_effective_total_amount_swapped
=== RUN   TestOrderbookUsecaseTestSuite/TestCreateFormattedLimitOrder/error_parsing_bid_unrealized_cancels
=== RUN   TestOrderbookUsecaseTestSuite/TestCreateFormattedLimitOrder/error_parsing_etas
=== RUN   TestOrderbookUsecaseTestSuite/TestCreateFormattedLimitOrder/error_converting_tick_to_price
=== RUN   TestOrderbookUsecaseTestSuite/TestCreateFormattedLimitOrder/error_parsing_placed_at
=== RUN   TestOrderbookUsecaseTestSuite/TestCreateFormattedLimitOrder/successful_order_processing
--- PASS: TestOrderbookUsecaseTestSuite (0.03s)
    --- PASS: TestOrderbookUsecaseTestSuite/TestCreateFormattedLimitOrder (0.03s)
        --- PASS: TestOrderbookUsecaseTestSuite/TestCreateFormattedLimitOrder/tick_not_found (0.00s)
        --- PASS: TestOrderbookUsecaseTestSuite/TestCreateFormattedLimitOrder/error_parsing_quantity (0.00s)
        --- PASS: TestOrderbookUsecaseTestSuite/TestCreateFormattedLimitOrder/error_parsing_placed_quantity (0.00s)
        --- PASS: TestOrderbookUsecaseTestSuite/TestCreateFormattedLimitOrder/error_getting_spot_price_scaling_factor (0.00s)
        --- PASS: TestOrderbookUsecaseTestSuite/TestCreateFormattedLimitOrder/error_parsing_bid_effective_total_amount_swapped (0.00s)
        --- PASS: TestOrderbookUsecaseTestSuite/TestCreateFormattedLimitOrder/error_parsing_bid_unrealized_cancels (0.00s)
        --- PASS: TestOrderbookUsecaseTestSuite/TestCreateFormattedLimitOrder/error_parsing_etas (0.00s)
        --- PASS: TestOrderbookUsecaseTestSuite/TestCreateFormattedLimitOrder/error_converting_tick_to_price (0.00s)
        --- PASS: TestOrderbookUsecaseTestSuite/TestCreateFormattedLimitOrder/error_parsing_placed_at (0.00s)
        --- PASS: TestOrderbookUsecaseTestSuite/TestCreateFormattedLimitOrder/successful_order_processing (0.00s)
PASS
ok      github.com/osmosis-labs/sqs/orderbook/usecase   0.090s
```

## Summary by CodeRabbit

- **New Features**
	- Introduced a mock implementation for the OrderBookRepository to enhance testing capabilities.
	- Added a public interface for creating formatted limit orders in the order book domain.
	- Implemented a comprehensive test suite for the OrderbookUsecase functionality, covering various scenarios for order creation.
	- Defined custom error types for improved error handling and reporting within the order book system.
	- Enhanced the OrderBookClient interface with new methods for fetching unrealized cancels and ticks in chunks.
	- Updated the LimitOrder struct to use higher precision decimal types for key fields.
	- Added a new endpoint to retrieve active orders associated with a specified Osmo wallet address.

- **Bug Fixes**
	- Added validation for `placedQuantity` in order creation to prevent invalid orders.
	- Adjusted return types for better performance and type handling in order book use case functions.

### CODE REVIEW

File: orderbook/usecase/orderbook_usecase_test.go [hunk 1]
```diff
@@ -0,0 +1,314 @@
...
+			usecase := orderbookusecase.New(
+				&orderbookRepo,
+				nil,
+				nil,
+				&tokensUsecase,
+				nil,
+			)
+
+			// Call the method under test
+			result, err := usecase.CreateFormattedLimitOrder(1, tc.order, tc.quoteAsset, tc.baseAsset, "someOrderbookAddress")
```
Comment (p0mvn) [inline orderbook/usecase/orderbook_usecase_test.go]: Let's define a constant `defaultPoolID`. Also, does it make sense to have at least one test with a different poolID and `orderbokAdddress` to ensure that the test does not break with default values.

File: orderbook/usecase/orderbook_usecase_test.go [hunk 2]
```diff
@@ -0,0 +1,314 @@
...
+		setupMocks    func(orderbookRepo *mocks.OrderbookRepositoryMock, tokensUsecase *mocks.TokensUsecaseMock)
+		expectedError string
+		expectedOrder orderbookdomain.LimitOrder
+	}{
+		{
+			name: "tick not found",
+			order: orderbookdomain.Order{
+				TickId: 99, // Non-existent tick ID
+			},
+			expectedError: "tick not found",
```
Comment (p0mvn) [inline orderbook/usecase/orderbook_usecase_test.go]: Instead of hardcoding error string, consider referencing an actual error returned from core codebase that we can define as a var and export via `export_test.go`

File: orderbook/usecase/orderbook_usecase_test.go [hunk 3]
```diff
@@ -0,0 +1,314 @@
...
+						},
+					}, true
+				}
+			},
+		},
+		{
+			name: "error getting spot price scaling factor",
+			order: orderbookdomain.Order{
+				Quantity:       "1000",
+				PlacedQuantity: "1500",
```
Comment (p0mvn) [inline orderbook/usecase/orderbook_usecase_test.go]: Consider reducing code duplication and hardcoding by defining constants

File: orderbook/usecase/orderbook_usecase_test.go [hunk 4]
```diff
@@ -0,0 +1,314 @@
...
+				Quantity: "invalid", // Invalid quantity
+			},
+			expectedError: "error parsing quantity",
+			setupMocks: func(orderbookRepo *mocks.OrderbookRepositoryMock, tokensUsecase *mocks.TokensUsecaseMock) {
+				orderbookRepo.GetTickByIDFunc = func(poolID uint64, tickID int64) (orderbookdomain.OrderbookTick, bool) {
+					return orderbookdomain.OrderbookTick{
+						Tick: &cosmwasmpool.OrderbookTick{
+							TickId: tickID,
+						},
+					}, true
```
Comment (p0mvn) [inline orderbook/usecase/orderbook_usecase_test.go]: The mocked `GetTickByIDFunc` seems repeating from the previous test. Would it be possible to define it on top? This would reduce code duplication, increasing maintainability and readability

File: orderbook/usecase/orderbook_usecase_test.go [hunk 5]
```diff
@@ -0,0 +1,314 @@
...
+							TickId: tickID,
+						},
+					}, false
+				}
+			},
+		},
+		{
+			name: "error parsing quantity",
+			order: orderbookdomain.Order{
+				Quantity: "invalid", // Invalid quantity
```
Comment (p0mvn) [inline orderbook/usecase/orderbook_usecase_test.go:643]: Could we also add a test where quantity is a large int that overflows and handled gracefully?

Thinking more about this, I think this is very likely to overflow and panic in prod. Could we instead do string to Dec conversion via `NewDecFromStr` while also adding a relevant test?

File: orderbook/usecase/orderbook_usecase_test.go [hunk 6]
```diff
@@ -0,0 +1,314 @@
...
+					return orderbookdomain.OrderbookTick{
+						Tick: &cosmwasmpool.OrderbookTick{
+							TickId: tickID,
+						},
+					}, true
+				}
+			},
+		},
+		{
+			name: "error parsing placed quantity",
```
Comment (p0mvn) [inline orderbook/usecase/orderbook_usecase_test.go:666]: I think that we need to change the way we parse quantity for placed quantity as well `NewDecFromStr`

File: orderbook/usecase/orderbook_usecase_test.go [hunk 7]
```diff
@@ -0,0 +1,314 @@
...
+			order: orderbookdomain.Order{
+				Quantity:       "1000",
+				PlacedQuantity: "invalid", // Invalid placed quantity
+			},
+			expectedError: "error parsing placed quantity",
```
Comment (p0mvn) [inline orderbook/usecase/orderbook_usecase_test.go]: We should have a test where placed quantity is zero. Currently, the system will panic due to division by zero, and we need to add a guarding check against the panic

File: orderbook/usecase/orderbook_usecase_test.go [hunk 8]
```diff
@@ -0,0 +1,314 @@
...
+			order: orderbookdomain.Order{
+				TickId:         1,
+				OrderId:        1,
+				OrderDirection: "bid",
+				Owner:          "owner1",
+				Quantity:       "1000",
+				PlacedQuantity: "1500",
+				Etas:           "500",
+				ClaimBounty:    "10",
+				PlacedAt:       "1634764800000",
```
Comment (p0mvn) [inline orderbook/usecase/orderbook_usecase_test.go]: Let's review and revalidate all numerical values against overflows

**Review by p0mvn — COMMENTED** [review]
Nice work. Left some comments - please take a look.

Mostly, recommendations around overflows and reducing code duplication

File: orderbook/usecase/orderbook_usecase_test.go (hunk 1 above)
Comment (deividaspetraitis) [reply orderbook/usecase/orderbook_usecase_test.go]: PoolID is passed because it is requires as argument, actual function that uses this argument is mocked, thus it can be any number value. To avoid use of constant I have moved it to `tc.poolID` that defauts to `0` at this moment for all test cases.

File: orderbook/usecase/orderbook_usecase_test.go (hunk 2 above)
Comment (deividaspetraitis) [reply orderbook/usecase/orderbook_usecase_test.go]: That is a valid point, my initial attempt implementing that failed, since for this method client(s) specific errors are irrelevant, and added complexity by defining errors and introduced additional code outweights the benefts, thus I would leave as it is at least for now. Let me know if you still think we should avoid doing that.

File: orderbook/usecase/orderbook_usecase_test.go (hunk 3 above)
Comment (deividaspetraitis) [reply orderbook/usecase/orderbook_usecase_test.go]: The main reason I am not reusing contants here was to make values local to the test case to avoid adding cognitive load. Let me know if that makes sense.

File: orderbook/usecase/orderbook_usecase_test.go (hunk 4 above)
Comment (deividaspetraitis) [reply orderbook/usecase/orderbook_usecase_test.go]: Good point! I have just did that.

File: orderbook/usecase/orderbook_usecase_test.go (hunk 5 above)
Comment (deividaspetraitis) [reply orderbook/usecase/orderbook_usecase_test.go:643]: Good catch! Added overflow checks. :rocket:

File: orderbook/usecase/orderbook_usecase_test.go (hunk 7 above)
Comment (deividaspetraitis) [reply orderbook/usecase/orderbook_usecase_test.go]: Good catch! Added test to cover this case. :+1:

File: orderbook/usecase/orderbook_usecase_test.go (hunk 6 above)
Comment (deividaspetraitis) [reply orderbook/usecase/orderbook_usecase_test.go:666]: We do [return](https://github.com/osmosis-labs/sqs/blob/e4ce75541f31c4de7baf6da3e8b95fae6744f03e/orderbook/usecase/orderbook_usecase.go#L383) quantity as `int64` for the clients, if for example `quantity` overflows we will catch it with `strconv.ParseInt`. At this point I am trying to have exactly same precision logic as FE had. If we want to have some adjusments regarding that, maybe we can create a separate task to track that?

File: orderbook/usecase/orderbook_usecase_test.go (hunk 8 above)
Comment (deividaspetraitis) [reply orderbook/usecase/orderbook_usecase_test.go]: Good point, implemented! :+1:

File: orderbook/usecase/orderbook_usecase_test.go (hunk 2 above)
Comment (p0mvn) [reply orderbook/usecase/orderbook_usecase_test.go]: I still think it is worth it doing something like [this](https://github.com/osmosis-labs/sqs/blob/f1740c12a07ac7a2c9df284d45b1323224a755ef/domain/errors.go#L318) so that if an error message is ever updated, we only need to update it once and don't worry about several hardcoded tests

File: orderbook/usecase/orderbook_usecase_test.go (hunk 6 above)
Comment (p0mvn) [reply orderbook/usecase/orderbook_usecase_test.go:666]: My point is that it will frequently overflow in `strconv.ParseInt`, making the system unusable.

Consider 1000 DAI which is 1000 * 10**18 in base units.

```
>>> 2**63 - 1 - 1000 * 10**18
-990776627963145224193
```

Just $1000 of DAI overflows here so we should use the appropriate chain types to prevent that.

Comment (deividaspetraitis) [reply orderbook/usecase/orderbook_usecase_test.go:666]: Ah now I see you point, thanks for elaboration! :pray: The idea is that currently we support only stables ( `USDC`, `USDT` ) as a quotes each of which has `6` decimal places,  meaning that paid amount should be larger than `1*10**12` in order to overflow:

```
>>> 2**63 - 1 - 1*10**13 * 10**6
-776627963145224193
```

In order to support non stables as a quotes would require some adjustments on FE side to, meaning we can be reasurred that somebody from product will not enable support of more asset types without anyone noticing from engineering team ( including us ) in advance.

Having this information, do you think we still should go and implement this change within scope of this task?

File: orderbook/usecase/orderbook_usecase_test.go (hunk 2 above)
Comment (deividaspetraitis) [reply orderbook/usecase/orderbook_usecase_test.go]: Thanks, I have implemented as per your suggestion [here](https://github.com/osmosis-labs/sqs/pull/492/files#diff-a585202576c1e374948b8b760a1fe96534e5a35a29ccb3e846e5a7548c714f0a). Would that work? :pray:

File: orderbook/usecase/orderbook_usecase_test.go (hunk 6 above)
Comment (p0mvn) [reply orderbook/usecase/orderbook_usecase_test.go:666]: I'm not sure I understand how stables as quotes excludes the possibility of overflow.

An order can be placed in both base and quote denominations so amounts are not limited to only the listed stables precision. 

Additionally, there are 18 decimal stables like DAI which we can trivially add support for.

I would vote to do it right now so that we don't need to worry about it. I also have low confidence in our cross-team communication on breaking changes so being extra diligent around such edge cases would pay-off down the road, helping to avoid incidents.

File: orderbook/usecase/orderbook_usecase_test.go (hunk 2 above)
Comment (p0mvn) [reply orderbook/usecase/orderbook_usecase_test.go]: yes, nice job!

File: orderbook/usecase/orderbook_usecase_test.go (hunk 6 above)
Comment (deividaspetraitis) [reply orderbook/usecase/orderbook_usecase_test.go:666]: Thanks! I have implemented your suggested improvement [here](https://github.com/osmosis-labs/sqs/pull/492/files#diff-ae675fb65861261e203f8bf1a1aacfd68815f3cb763d28db817797e9a88904b3R278) and have adjusted tests to check cases for possible `osmomath.Dec` overflows. Would that work?

File: orderbook/usecase/orderbook_usecase.go [hunk 9]
```diff
@@ -262,94 +266,124 @@ func (o *orderbookUseCaseImpl) createFormattedLimitOrder(
...
+	placedQuantity, err := osmomath.NewDecFromStr(order.PlacedQuantity)
 	if err != nil {
-		return orderbookdomain.LimitOrder{}, fmt.Errorf("error parsing placed quantity: %w", err)
+		return orderbookdomain.LimitOrder{}, types.ParsingPlacedQuantityError{
+			PlacedQuantity: order.PlacedQuantity,
+			Err:            err,
+		}
+	}
+
+	if zero := osmomath.NewDec(0); placedQuantity.Equal(zero) || placedQuantity.LT(zero) {
```
Comment (p0mvn) [inline orderbook/usecase/orderbook_usecase.go]: nit: we could define `zero := osmomath.NewDec(0)` in the global space so that it does not have to be re-initilized for every client call

File: orderbook/usecase/orderbook_usecase.go [hunk 10]
```diff
@@ -262,94 +266,124 @@ func (o *orderbookUseCaseImpl) createFormattedLimitOrder(
...
+		}
 	}
 
 	// Determine tick values and unrealized cancels based on order direction
 	var tickEtas, tickUnrealizedCancelled int64
 	if order.OrderDirection == "bid" {
 		tickEtas, err = strconv.ParseInt(tickState.BidValues.EffectiveTotalAmountSwapped, 10, 64)
 		if err != nil {
-			return orderbookdomain.LimitOrder{}, fmt.Errorf("error parsing bid effective total amount swapped: %w", err)
+			return orderbookdomain.LimitOrder{}, types.ParsingTickValuesError{
```
Comment (p0mvn) [inline orderbook/usecase/orderbook_usecase.go:332]: I think we might also change the parsing for tick ETAs.

Synched with Connor on definition:
> There's two places where we record an effective total amount swapped value. One is in the tick (for each direction) and one is in the order itself. When you place an order the current tick's ETAS (effective total amount swapped) value is recorded under the order's etas field. As swaps are performed on the tick we update the tick's ETAS value (for the given direction). The tick's ETAS value is then compared against the order's current etas field to determine how much of the order has been filled. If the order is partially filled we update the etas field to be the current tick's ETAS, if the order is fully filled it's simply removed.

File: orderbook/usecase/orderbook_usecase_test.go [hunk 11]
```diff
@@ -0,0 +1,542 @@
...
+			usecase := orderbookusecase.New(
+				&orderbookRepo,
+				nil,
+				nil,
+				&tokensUsecase,
+				nil,
+			)
+
+			// Call the method under test
+			result, err := usecase.CreateFormattedLimitOrder(tc.poolID, tc.order, tc.quoteAsset, tc.baseAsset, "someOrderbookAddress")
```
Comment (p0mvn) [inline orderbook/usecase/orderbook_usecase_test.go]: nit: would be good to have a test case with the address varied for full coverage of the input set

**Review by p0mvn — COMMENTED** [review]
One blocking comment remaining:
- The concern around ETAs overflow

File: orderbook/usecase/orderbook_usecase.go (hunk 9 above)
Comment (deividaspetraitis) [reply orderbook/usecase/orderbook_usecase.go]: Addressed! :+1:

File: orderbook/usecase/orderbook_usecase.go (hunk 10 above)
Comment (deividaspetraitis) [reply orderbook/usecase/orderbook_usecase.go:332]: Oh, that's a really good catch, thanks! Updated as per your suggestion. :pray:

**Review by p0mvn — APPROVED** [review]
LGTM

Could we address this one dangling comment pre-merge?
https://github.com/osmosis-labs/sqs/pull/492#discussion_r1742422670

File: orderbook/usecase/orderbook_usecase_test.go (hunk 11 above)
Comment (deividaspetraitis) [reply orderbook/usecase/orderbook_usecase_test.go]: Addressed with [4e1e945](https://github.com/osmosis-labs/sqs/pull/492/commits/4e1e9457636004e1f5bf701246a10526abad37a4) :+1:
